src/java.base/share/classes/java/lang/StringLatin1.java

```diff
@@ -388,43 +388,94 @@ public static boolean regionMatchesCI(byte[] value, int toffset,
-            char u2 = Character.toUpperCase(c2);
-            if (u1 == u2) {
-                continue;
-            }
-            if (Character.toLowerCase(u1) == Character.toLowerCase(u2)) {
+            if (c1 == c2 || StringUTF16.latin1EqualsIgnoreCase(c1, c2) || turkicIEqualsIgnoreCase(c1, c2)) {
                 continue;
             }
             return false;
         }
         return true;
     }
 
+    /**
+     * In ISO/IEC 8859-1, the case-folding letters are found in the
+     * following three code point range pairs:
+     *
+     * 0x41-0x5A: Uppercase ASCII letters (A-Z)
+     * 0x61-0x7A: Lowercase ASCII letters (a-z)
+     *
+     * 0xC0-0xD6: Uppercase non-ASCII letters, range 1 (A-GRAVE - O with Diaeresis)
+     * 0xE0-0xF6: Lowercase non-ASCII letters, range 1 (a-grave - o with Diaeresis)
+     *
+     * 0xD8-0xDE: Uppercase non-ASCII letters, range 2 (O with slash - Thorn)
+     * 0xF8-0xFE: Lowercase non-ASCII letters, range 2 (o with slash - thorn)
+     *
+     * While both ASCII letter ranges are contiguous, the non-ASCII letters are not:
+     *
+     * The non-ASCII uppercase range is split in two by the multiplication sign 0xD7.
+     * The non-ASCII lowercase range is split in two by the division sign 0xF7.
+     *
+     * The relative distance between each lowercase and uppercase case-folding letter
+     * is always 32 (0x20), where the uppercase code point is always lowest. The division
+     * and multiplication code points have the same relationship.
+     *
+     * Since 0x20 is a single bit, we can apply the 'oldest ASCII trick in the book':
+     *
+     * 'C' | 0x20 == 'c'
+     *
+     * The reverse is also true:
+     *
+     * 'c' & 0xDF = 'C'
+     *
+     * @param a byte encoded in ISO/IEC 8859-1
+     * @param b another byte encoded in ISO/IEC 8859-1
+     * @return true if the two bytes are considered equals ignoring case in ISO/IEC 8859-1.
+     */
+    private static boolean latin1EqualsIgnoreCase(byte a, byte b) {
+        int A = a & 0xDF; // uppercase a
+        return ( (A >= 'A' && A <= 'Z') // A in range A-Z
+                || (A >= 0xC0 && A <= 0xDE)) // ..or Agrave-Thorn
+                && A != 0xD7 // But not multiply
+                && A == (b & 0xDF); // b has same uppercase
+    }
+
+    /**
+     * Latin1 includes two 'i' code points:
+     *
+     * 0x49: 'Latin Capital Letter I'
+     * 0x69: 'Latin Small Letter I'
+     *
+     * Unicode adds two 'i' code points for Turkic languages:
+     *
+     * 0x130: 'Latin Capital Letter I with Dot Above'
+     * 0x131: 'Latin Small Letter Dotless I'
+     *
+     * These four code points should be considered equals when ignoring case.
+     *
+     * @param lat a latin1 code point
+     * @param uc a unicode code point
+     * @return true if the latin code point is either 'i' or 'I'
+     * and the unicode code point is either 0x130 or 0x131
+      */
+    private static boolean turkicIEqualsIgnoreCase(char lat, char uc) {
```

Comment:
turkic -> turkish?

Reply:
I believe "turkic" is technically correct. Turkish is the most widely spoken language of the family of turkic langages, but not the only one.

At any rate, this code is obsolete since it turned out that there are several Unicode code points which fold into latin1, so we need a more general solution. See comment from Claes above.

Reply:
I'm no expert but afaict several turkic languages, including azeri and kazakh, have adopted the modern turkish latin-based script that include these particular glyphs. Turkic might be the more inclusive term here.